lib/phoenix_component/macro_component.ex

```diff
@@ -0,0 +1,277 @@
+defmodule Phoenix.Component.MacroComponent do
+  @moduledoc """
+  A macro component is a special type of component that can modify its content
+  at compile time.
+
+  Instead of introducing a special tag syntax like `<#macro-component>`, LiveView
+  decided to "hide" macro components behind a special `:type` attribute. The reason
+  for this is that most of the time, the job of the macro component is to take its
+  content and extract it so somewhere else, for example to a file in the local
+  file system. A good example for this is `Phoenix.LiveView.ColocatedHook` and
+  `Phoenix.LiveView.ColocatedJS`.
+
+  ## AST
+
+  Macro components work by defining a callback module that implements the
+  `Phoenix.LiveView.MacroComponent` behaviour. The module's `c:transform/2` callback
+  is called for each macro component used while LiveView compiles a HEEx component:
+
+  ```heex
+  <div id="hey" phx-hook=".foo">
+    <!-- content -->
+  </div>
+
+  <script :type={ColocatedHook} name=".foo">
+    export default {
+      mounted() {
+        this.el.firstElementChild.textContent = "Hello from JS!"
+      }
+    }
+  </script>
+  ```
+
+  In this example, the `ColocatedHook`'s `c:transform/2` callback will be invoked
+  with the AST of the `<script>` tag:
+
+  ```elixir
+  {"script",
+    [{":type", {:__aliases__, [line: 1], [:ColocatedHook]}}, {"name", ".foo"}],
+    [
+      "\\n  export default {\\n    mounted() {\\n      this.el.firstElementChild.textContent = \\"Hello from JS!\\"\\n    }\\n  }\\n"
+    ]}
+  ```
+
+  The `Phoenix.Component.MacroComponent.AST` module provides some utilities to work with the AST,
+  but in general it is quite simple:
```

Comment:
Generally speaking, it is good to avoid words like "simple", "just" or "easy" in docs, because if someone doesn't get it, then that doesn't sound simple, and if someone immediately gets it, then they will see it is simple anyway. :)